plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php

```diff
@@ -0,0 +1,1203 @@
+use Automattic\WooCommerce\Blocks\BlockTypes\PriceFilter;
+use Automattic\WooCommerce\Blocks\BlockTypes\RatingFilter;
+use Automattic\WooCommerce\Blocks\BlockTypes\StockFilter;
+use WP_Query;
+use WC_Tax;
+
+/**
+ * ProductQueryBuilder class.
+ * Responsible for constructing and modifying product queries.
+ */
+class ProductQueryBuilder {
+
+	/**
+	 * All query args from WP_Query.
+	 *
+	 * @var array
+	 */
+	protected $valid_query_vars;
+
+	/**
+	 * Orderby options not natively supported by WordPress REST API
+	 *
+	 * @var array
+	 */
+	protected $custom_order_opts = array( 'popularity', 'rating', 'post__in', 'price', 'sales', 'menu_order', 'random' );
+
+	/**
+	 * All the query args related to the filter by attributes block.
+	 *
+	 * @var array
+	 */
+	protected $attributes_filter_query_args = array();
+
+	/**
+	 * Collection handler store.
+	 *
+	 * @var array
+	 */
+	protected $collection_handler_store = array();
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->valid_query_vars = $this->get_valid_query_vars();
+		add_filter( 'posts_clauses', array( $this, 'add_price_range_filter_posts_clauses' ), 10, 2 );
+	}
+
+	/**
+	 * Set the collection handler store.
+	 *
+	 * @param array $collection_handler_store The collection handler store containing registered collection handlers.
+	 */
+	public function set_collection_handler_store( $collection_handler_store ) {
+		$this->collection_handler_store = $collection_handler_store;
+	}
+
+	/**
+	 * Set collection handler.
+	 *
+	 * @param string $collection_name The name of the custom collection.
+	 * @param array  $handlers        Collection handlers.
+	 */
+	public function set_collection_handler( $collection_name, $handlers ) {
+		$this->collection_handler_store[ $collection_name ] = $handlers;
+	}
+
+	/**
+	 * Set attributes filter query args.
+	 *
+	 * @param array $args The attributes filter query arguments.
+	 */
+	public function set_attributes_filter_query_args( $args ) {
+		$this->attributes_filter_query_args = $args;
+	}
+
+	/**
+	 * Return or initialize $valid_query_vars.
+	 *
+	 * @return array
+	 */
+	private function get_valid_query_vars() {
+		if ( ! empty( $this->valid_query_vars ) ) {
+			return $this->valid_query_vars;
+		}
+
+		$valid_query_vars       = array_keys( ( new WP_Query() )->fill_query_vars( array() ) );
+		$this->valid_query_vars = array_merge(
+			$valid_query_vars,
+			// fill_query_vars doesn't include these vars so we need to add them manually.
+			array(
+				'date_query',
+				'exact',
+				'ignore_sticky_posts',
+				'lazy_load_term_meta',
+				'meta_compare_key',
+				'meta_compare',
+				'meta_query',
+				'meta_type_key',
+				'meta_type',
+				'nopaging',
+				'offset',
+				'order',
+				'orderby',
+				'page',
+				'post_type',
+				'posts_per_page',
+				'suppress_filters',
+				'tax_query',
+				'isProductCollection',
+				'priceRange',
+			)
+		);
+
+		return $this->valid_query_vars;
+    }
```

Comment:
I am seeing a lint error here. Are you able to see it? Or it's just me 😅
![image](https://github.com/user-attachments/assets/dfe1ad62-a488-4ccc-aff3-f3379d6fe02f)